src/adapter/tracer/ptrace.rs

```diff
@@ -0,0 +1,468 @@
+        for (i, lim) in BOUNDARIES.iter().enumerate() {
+            if num <= *lim {
+                self.buckets[i] = self.buckets[i].saturating_add(1);
+                break;
+            }
+        }
+    }
+
+    /// Returns the largest value that will be accepted into each bucket.
+    pub fn boundaries(&self) -> &'static [u64; 16] {
+        &BOUNDARIES
+    }
+
+    /// Returns the number of values stored in the histogram.
+    pub fn count(&self) -> u32 {
+        self.buckets()
+            .iter()
+            .copied()
+            .fold(0, |acc, num| acc.saturating_add(num))
+    }
+
+    /// Returns the count of each bucket
+    pub fn buckets(&self) -> &[u32; 16] {
+        &self.buckets
+    }
+}
+
+impl Default for ExpHistogram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A summary of key timing statistics.
+#[derive(Debug, Clone)]
+pub struct Summary {
+    hist: ExpHistogram,
+    min: Duration,
+    max: Duration,
+    sum: Duration,
+}
+
+impl Summary {
+    // By initialising with a duration, we don't require min/max to be options.
+    pub fn new(duration: Duration) -> Summary {
+        let mut hist = ExpHistogram::new();
+        hist.add(duration.as_nanos() as u64);
+
+        Summary {
+            hist,
+            min: duration,
+            max: duration,
+            sum: duration,
+        }
+    }
+
+    /// Add a new time to the summary.
+    pub fn update(&mut self, duration: Duration) {
+        self.hist.add(duration.as_nanos() as u64);
+
+        self.min = self.min.min(duration);
+        self.max = self.max.max(duration);
+        self.sum = self.sum.saturating_add(duration);
+    }
+
+    /// Returns the number of items that have been processed.
+    pub fn count(&self) -> u32 {
+        self.hist.count()
+    }
+
+    /// Returns the total time
+    pub fn total(&self) -> Duration {
+        self.sum
+    }
+
+    /// Returns the fastest operation
+    pub fn min(&self) -> Duration {
+        self.min
+    }
+
+    /// Returns the slowest operation
+    pub fn max(&self) -> Duration {
+        self.max
+    }
+
+    /// Returns the exponential histogram with times in nanoseconds.
+    pub fn histogram(&self) -> &ExpHistogram {
+        &self.hist
+    }
+
+    /// Returns the mean duration.
+    pub fn mean(&self) -> Duration {
+        // This cannot panic since count() is always at least 1.
+        self.sum / self.count()
+    }
+}
+
+/// Records and stores operations performed by the adapter.
+///
+/// This struct is intended for use inside of a TracingAdapter.
+/// Operations must be recorded sequentially in chronological order.
+/// Recording out-of-order operations will lead to invalid state.
+#[derive(Debug, Clone)]
+pub struct Tracer {
+    pub calls: BTreeMap<FunctionCall, Summary>,
+
+    // When we measure the time of an iterator, we also measure the time spent
+    // evaluating its inputs. We must therefore subtract this time out when
```

Comment:
nit: possible confusion with "time out"
```suggestion
    // evaluating its inputs. We must therefore subtract out this time when
```